design/design.md

```diff
@@ -0,0 +1,414 @@
+For Mithril, if the data elements are signatures, the weight oracle would
+have to report for any signature the amount of stake of the creator of the
+signature.
+Clearly, this is impossible.
+Even if the data element was a pair of public key and signature, the weight
+oracle would have no way of verifying that the public key owns any stake.
+Again, research is needed to construct the right ALBA scheme on
+paper before implementing it.
+
+### Correctness
+
+The definitions of ALBA require, for input parameters $\lambda_{sec}$ and
+$\lambda_{rel}$, that the soundness error is at most $2^{-\lambda_{sec}}$
+and the completeness error is at most $2^{-\lambda_{rel}}$.
+Ideally, our implementation should also make this guarantee.
+
+However, at this point, this is not our goal, since achieving that would
+require significant additional effort.
+We need to roll out a reference implemenentation so that our primary clients
+(Mithril, Peras, Leios) can incorporate it in their prototypes, and also so
+that the external community can study and experiment with ALBA as well.
+
+The following is a list of known discrepancies that prevent us from making
+the rigorous guarantees about soundness and completeness errors.
+1. tiny rounding errors in parameter derivation due to the use of
+floating-point numbers
+2. tiny errors when using a uniformly random hash output to sample distributions
+needed for ALBA such as Bernoulli, uniform, binomial, Poisson
+3. lack of a rigorous analysis of the $H_1$ optimization described in
+section 3.3 of the paper
+
+### Performance
+
+There are three performance metrics in consideration:
+1. proof size;
+2. communication complexity;
+3. computational complexity.
+
+The space complexity (memory usage) is not considered since it is very small
+in any ALBA scheme.
+
+Our three main applications, Mithril, Peras and Leios, will be offered a
+tradeoff between proof size and communication complexity.
+
+Peras requires the ALBA proof to fit in a single block.
+It is only needed if Peras goes into cooldown which we should expect to
+happen rarely.
+Depending on how ALBA is used, the lower bound parameter $n_f$ will be set
+to either 50% or 75% of the total stake.
+Leios will produce ALBA proofs very frequently.
+Their requirement on proof size -- below 5 kilobytes.
+Mithril currently uses a lottery;
+therefore, using ALBA instead can only decrease the proof size.
+It is possible that Peras and Leios requirements on the proof size cannot
+be satisfied using ALBA alone.
+In that case, further techniques can be considered such as SNARKs and
+taking the stake distribution into account (todo: see fait-accompli).
+
+These protocols' requirements about communication and computational complexity
+are not currently known.
+However, the communication complexity of the implemented ALBA schemes will not greatly
+exceed that of the simple lottery, and the prover running time is expected to be
+in the order of a second.
+
+### Interfaces
```

Comment:
Stop talking about Peras and Leios. You can just write that we will implement bindings